### PR TITLE
Update Pytorch Lightning Datamodule

### PR DESCRIPTION
**Pull request by wiederm**
## Description
This PR updates the pl Datamodule to enable multi node training. It also simplifies and standardizes the use of the AtomicSelfEnergy class that is used in the DataModule and Potential to keep track of self energies. It also standardizes the use of the Datamodule instead of the TorchDataset wherever possible. `preapare_data()` and `setup()` have to be called explicitly.

Before, we implemented a separate class for each Dataset and initialized it like this:

```python
from modelforge.dataset.qm9 import QM9Dataset
from modelforge.dataset.utils import FirstComeFirstServeSplittingStrategy
from modelforge.dataset.dataset import TorchDataModule

data = QM9Dataset(for_unit_testing=True)
dataset = TorchDataModule(
    data, batch_size=32, splitting_strategy=FirstComeFirstServeSplittingStrategy()
)
dataset.prepare_data(
    remove_self_energies=False, normalize=False, regression_ase=False
)
```
Using the new structure the `Trainer` calls `prepare_data` and `setup`. If it is used without a `Trainer` the user has to call this manually.

```python
dm = DataModule(
    name=datasets_to_test.name,
    batch_size=512,
    splitting_strategy=FirstComeFirstServeSplittingStrategy(),
    normalize=False,
    remove_self_energies=False,
    for_unit_testing=True,
)

# NOTE: only necessary if NOT passed to a trainer
dm.prepare_data()
dm.setup()
````
All control parameters for the dataset and dataloader are now exposed at a central interface (the `init` method).

It implements the `prepare_data` and `setup` methods of the dataset as recommended in the [documentation](https://lightning.ai/docs/pytorch/stable/data/datamodule.html). This had some downstream effects and the PR is larger than initially anticipated.

## Todos
  - [x] refactor `prepare_data` and `setup` methods
  - [x] refactor the logic for passing the atomic self-energies to the DataModule (and the potential)
  - [x] updated tests with the new data generation structure
  - [x] Refactored `Loss` class to avoid recalculation of energy when force is requested. Also, the Loss class is now independent of the actual loss function.

## Status
- [x] Ready to go

### CODE REVIEW

**Review by chrisiacovella — APPROVED** [review]
A few small comments. This is a good change, should be cleaner.

File: modelforge/dataset/dataset.py [hunk 1]
```diff
@@ -736,47 +736,178 @@ def create_dataset(
...
+        self,
+    ) -> None:
+        """
+        Prepares the dataset for use. This method is responsible for the initial processing of
+        the data such as calculating self energies, normalizing, and splitting. It is executed
+        only once per node.
+        """
+        from modelforge.dataset import _IMPLEMENTED_DATASETS
+
+        dataset_class = _IMPLEMENTED_DATASETS[self.name]
```
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/dataset/dataset.py:799]: I agree! We can use either `Enum` or `dataclass`/`NamedTuple`?

File: modelforge/dataset/dataset.py [hunk 2]
```diff
@@ -736,47 +736,178 @@ def create_dataset(
...
+            - self_energies are None, self._ase will be used
+            - regression_ase is True, self_energies will be calculated
+        """
         super().__init__()
-        self.data = data
+
+        self.name = name
         self.batch_size = batch_size
-        self.split_idxs = None
+        self.splitting_strategy = splitting_strategy or RandomRecordSplittingStrategy()
```
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/dataset/dataset.py:775]: sure! reverted this!

File: modelforge/dataset/dataset.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/dataset/dataset.py]: decided for `Enum`
